File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/businesses/MyBusinessesViewModel.kt
@file:Suppress("MagicNumber")

package app.pantopus.android.ui.screens.businesses

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.businesses.BusinessMembership
import app.pantopus.android.data.api.models.businesses.BusinessProfileDto
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.businesses.BusinessesRepository
import app.pantopus.android.ui.screens.shared.list_of_rows.AvatarBackground
import app.pantopus.android.ui.screens.shared.list_of_rows.AvatarBadgeSize
import app.pantopus.android.ui.screens.shared.list_of_rows.BannerConfig
import app.pantopus.android.ui.screens.shared.list_of_rows.BannerCtaTint
import app.pantopus.android.ui.screens.shared.list_of_rows.GradientPair
import app.pantopus.android.ui.screens.shared.list_of_rows.ListOfRowsUiState
import app.pantopus.android.ui.screens.shared.list_of_rows.RowLeading
import app.pantopus.android.ui.screens.shared.list_of_rows.RowModel
import app.pantopus.android.ui.screens.shared.list_of_rows.RowSection
import app.pantopus.android.ui.screens.shared.list_of_rows.RowTemplate
import app.pantopus.android.ui.screens.shared.list_of_rows.RowTrailing
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * ViewModel for My businesses — wraps `GET /api/businesses/my-businesses`.
 *
 * T6.3f / P14 row anatomy (avatar-first, no tabs):
 *   leading  → 44dp business-violet avatar w/ verified badge when
 *              `profile.is_published == true`
 *   title    → business.name (or username fallback)
 *   subtitle → `<Category> · <Role>` (Owner / Manager / Staff)
 *   body     → "<city>, <state>" — or "Online only" when blank
 *   trailing → chevron (tap → business dashboard)
 *
 * Plus a business-tinted intro banner and a `.SecondaryCreate` FAB
 * tinted `FabTint.Business`.
 */
@HiltViewModel
class MyBusinessesViewModel
    @Inject
    constructor(
        private val repo: BusinessesRepository,
    ) : ViewModel() {
        private val _state = MutableStateFlow<ListOfRowsUiState>(ListOfRowsUiState.Loading)
        val state: StateFlow<ListOfRowsUiState> = _state.asStateFlow()

        private val _banner = MutableStateFlow<BannerConfig?>(null)
        val banner: StateFlow<BannerConfig?> = _banner.asStateFlow()

        private var onOpenBusiness: (String) -> Unit = {}
        private var onRegister: () -> Unit = {}

        fun configureNavigation(
            onOpenBusiness: (String) -> Unit,
            onRegister: () -> Unit,
        ) {
            this.onOpenBusiness = onOpenBusiness
            this.onRegister = onRegister
        }

        fun load() {
            if (_state.value is ListOfRowsUiState.Loaded) return
            refresh()
        }

        fun refresh() {
            _state.value = ListOfRowsUiState.Loading
            _banner.value = null
            viewModelScope.launch {
                when (val result = repo.myBusinesses()) {
                    is NetworkResult.Success -> applySuccess(result.data.businesses)
                    is NetworkResult.Failure -> _state.value = ListOfRowsUiState.Error(result.error.message)
                }
            }
        }

        private fun applySuccess(memberships: List<BusinessMembership>) {
            if (memberships.isEmpty()) {
                _state.value =
                    ListOfRowsUiState.Empty(
                        icon = PantopusIcon.Building2,
                        headline = "No businesses yet",
                        subcopy =
                            "Create a business profile to take quotes inside Pantopus and earn the violet verified mark.",
                        ctaTitle = "Register a business",
                        onCta = onRegister,
                    )
                return
            }
            val rows = memberships.map(::rowFor)
            _state.value =
                ListOfRowsUiState.Loaded(
                    sections = listOf(RowSection(id = "my-businesses", rows = rows)),
                    hasMore = false,
                )
            _banner.value =
                BannerConfig(
                    icon = PantopusIcon.Building2,
                    title = if (rows.size == 1) "1 verified business" else "${rows.size} verified businesses",
                    subtitle = "Tap any business to manage its inbox, gigs, and reviews",
                    tint = BannerCtaTint.Business,
                )
        }

        private fun rowFor(membership: BusinessMembership): RowModel {
            val title =
                membership.business.name?.takeIf { it.isNotEmpty() }
                    ?: membership.business.username?.takeIf { it.isNotEmpty() }
                    ?: "Untitled business"
            val category = categoryLabel(membership.profile)
            val role = roleLabel(membership.roleBase)
            val subtitle =
                listOfNotNull(category, role)
                    .joinToString(" · ")
                    .takeIf { it.isNotEmpty() }
            val locality =
                listOfNotNull(membership.business.city, membership.business.state)
                    .filter { it.isNotEmpty() }
                    .joinToString(", ")
                    .takeIf { it.isNotEmpty() }
            val body = locality ?: "Online only"
            val bodyIcon = if (locality != null) PantopusIcon.MapPin else PantopusIcon.Info

            return RowModel(
                id = membership.businessUserId,
                title = title,
                subtitle = subtitle,
                template = RowTemplate.AvatarKebab,
                leading =
                    RowLeading.AvatarWithBadge(
                        name = title,
                        imageUrl = membership.business.profilePictureUrl,
                        background =
                            AvatarBackground.Gradient(
                                GradientPair(
                                    start = PantopusColors.business,
                                    end = PantopusColors.business,
                                ),
                            ),
                        size = AvatarBadgeSize.Large,
                        verified = membership.profile?.isPublished == true,
                    ),
                trailing = RowTrailing.Chevron,
                onTap = { onOpenBusiness(membership.businessUserId) },
                body = body,
                bodyIcon = bodyIcon,
            )
        }

        private fun categoryLabel(profile: BusinessProfileDto?): String? {
            val first = profile?.categories?.firstOrNull()?.takeIf { it.isNotEmpty() }
            if (first != null) return first.replace('_', ' ').replaceFirstChar { it.uppercase() }
            val bizType = profile?.businessType?.takeIf { it.isNotEmpty() }
            return bizType?.replace('_', ' ')?.replaceFirstChar { it.uppercase() }
        }

        private fun roleLabel(roleBase: String?): String? =
            when (roleBase) {
                "owner" -> "Owner"
                "admin" -> "Admin"
                "manager" -> "Manager"
                "staff" -> "Staff"
                "viewer" -> "Viewer"
                "editor" -> "Editor"
                null, "" -> null
                else -> roleBase.replaceFirstChar { it.uppercase() }
            }
    }

File: frontend/apps/ios/Pantopus/Features/Businesses/MyBusinessesViewModel.swift
//
//  MyBusinessesViewModel.swift
//  Pantopus
//
//  T6.3f / P14 — the user's index of every verified business they own
//  or staff. Backed by `GET /api/businesses/my-businesses`.
//
//  Row shape (avatar-first, no tabs):
//    leading  → 56pt category-tinted logo avatar (initials over
//               business-violet OR category-flavoured gradient)
//    title    → business name
//    subtitle → `<Category> · <Role>` (Owner / Manager / Staff)
//    body     → locality or "Online only" when no city/state
//    trailing → chevron (tap → business dashboard)
//
//  FAB: 60pt secondaryCreate, .business tint — pushes to a register-
//  business placeholder until the compose flow lands.
//

import Foundation
import Observation
import SwiftUI

@Observable
@MainActor
final class MyBusinessesViewModel: ListOfRowsDataSource {
    let title = "My businesses"
    var topBarAction: TopBarAction? {
        nil
    }

    let tabs: [ListOfRowsTab] = []
    var selectedTab: String = ""

    var fab: FABAction? {
        FABAction(
            icon: .building2,
            accessibilityLabel: "Register a business",
            variant: .secondaryCreate,
            tint: .business,
            handler: onRegister
        )
    }

    var banner: BannerConfig? {
        guard case let .loaded(sections, _) = state,
              let count = sections.first?.rows.count,
              count > 0
        else {
            return nil
        }
        return BannerConfig(
            icon: .building2,
            title: count == 1 ? "1 verified business" : "\(count) verified businesses",
            subtitle: "Tap any business to manage its inbox, gigs, and reviews",
            tint: .business
        )
    }

    private(set) var state: ListOfRowsState = .loading

    private let api: APIClient
    private let onOpenBusiness: (String) -> Void
    private let onRegister: @Sendable () -> Void

    init(
        api: APIClient = .shared,
        onOpenBusiness: @escaping (String) -> Void = { _ in },
        onRegister: @escaping @Sendable () -> Void = {}
    ) {
        self.api = api
        self.onOpenBusiness = onOpenBusiness
        self.onRegister = onRegister
    }

    func load() async {
        if case .loaded = state { return }
        state = .loading
        await fetch()
    }

    func refresh() async {
        await fetch()
    }

    func loadMoreIfNeeded() async {}

    private func fetch() async {
        do {
            let response: MyBusinessesResponse = try await api.request(
                BusinessesEndpoints.myBusinesses()
            )
            let rows = response.businesses.map { row(for: $0) }
            if rows.isEmpty {
                state = .empty(
                    ListOfRowsState.EmptyContent(
                        icon: .building2,
                        headline: "No businesses yet",
                        subcopy: "Create a business profile to take quotes inside Pantopus and earn the violet verified mark.",
                        ctaTitle: "Register a business",
                        onCTA: onRegister
                    )
                )
            } else {
                state = .loaded(sections: [RowSection(rows: rows)], hasMore: false)
            }
        } catch {
            state = .error(message: (error as? APIError)?.errorDescription ?? "Something went wrong.")
        }
    }

    private func row(for membership: BusinessMembership) -> RowModel {
        let business = membership.business
        let displayTitle = business.name?.nilIfEmpty
            ?? business.username?.nilIfEmpty
            ?? "Untitled business"
        let category = categoryLabel(for: membership.profile)
        let role = roleLabel(for: membership.roleBase)
        let subtitleParts = [category, role].compactMap { $0 }
        let subtitle = subtitleParts.isEmpty ? nil : subtitleParts.joined(separator: " · ")

        let locality = [business.city, business.state]
            .compactMap { $0?.nilIfEmpty }
            .joined(separator: ", ")
            .nilIfEmpty
        let body = locality ?? "Online only"

        return RowModel(
            id: membership.businessUserId,
            title: displayTitle,
            subtitle: subtitle,
            template: .avatarKebab,
            leading: .avatarWithBadge(
                name: displayTitle,
                imageURL: business.profilePictureURL.flatMap(URL.init(string:)),
                background: .gradient(GradientPair(
                    start: Theme.Color.business,
                    end: Theme.Color.business
                )),
                size: .large,
                verified: membership.profile?.isPublished == true
            ),
            trailing: .chevron,
            onTap: { @Sendable in
                Task { @MainActor in self.onOpenBusiness(membership.businessUserId) }
            },
            body: body,
            bodyIcon: locality == nil ? .info : .mapPin
        )
    }

    private func categoryLabel(for profile: BusinessProfileDTO?) -> String? {
        if let first = profile?.categories?.first, !first.isEmpty {
            return first.replacingOccurrences(of: "_", with: " ").capitalized
        }
        if let type = profile?.businessType, !type.isEmpty {
            return type.replacingOccurrences(of: "_", with: " ").capitalized
        }
        return nil
    }

    private func roleLabel(for roleBase: String?) -> String? {
        guard let roleBase, !roleBase.isEmpty else { return nil }
        switch roleBase {
        case "owner": return "Owner"
        case "admin": return "Admin"
        case "manager": return "Manager"
        case "staff": return "Staff"
        case "viewer": return "Viewer"
        case "editor": return "Editor"
        default: return roleBase.capitalized
        }
    }
}

private extension String {
    var nilIfEmpty: String? {
        isEmpty ? nil : self
    }
}
